pkg/baseimg/base_image.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2022, 2023 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package baseimg
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"os"
+	"path"
+
+	"github.com/chainguard-dev/go-apk/pkg/apk"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/layout"
+	ocitypes "github.com/google/go-containerregistry/pkg/v1/types"
+
+	"chainguard.dev/apko/pkg/build/types"
+)
+
+type BaseImage struct {
+	img      v1.Image
+	apkIndex []byte
+	tmpDir   string
+	arch     types.Architecture
+}
+
+// See https://github.com/opencontainers/image-spec/blob/main/image-index.md#image-index-property-descriptions
+// Briefly: index.json can either list manifest of per arch images, or redirect to actual image index (nested case)
+func getUnnestedImageIndex(imgPath string) (v1.ImageIndex, error) {
+	index, err := layout.ImageIndexFromPath(imgPath)
+	if err != nil {
+		return nil, err
+	}
+	indexManifest, err := index.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+	for _, m := range indexManifest.Manifests {
+		if m.MediaType == ocitypes.OCIImageIndex {
+			return index.ImageIndex(m.Digest)
+		}
+	}
+	return index, nil
+}
+
+func getImageForArch(imgPath string, arch types.Architecture) (v1.Image, error) {
+	index, err := getUnnestedImageIndex(imgPath)
+	if err != nil {
+		return nil, err
+	}
+	indexManifest, err := index.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, m := range indexManifest.Manifests {
+		img, err := index.Image(m.Digest)
+		if err != nil {
+			return nil, err
+		}
+		config, err := img.ConfigFile()
+		if err != nil {
+			return nil, err
+		}
+		if config == nil {
+			return nil, fmt.Errorf("got image without config")
+		}
+		if config.Architecture == arch.ToOCIPlatform().Architecture {
+			return img, nil
+		}
+	}
+	return nil, fmt.Errorf("image for arch not found")
+}
+
+func New(imgPath string, apkIndexPath string, arch types.Architecture, tmpDir string) (*BaseImage, error) {
```

Comment:
Whether it's tmpDir or permanent dir it's a caller decision. Please call it another way.

Reply:
Done.